src/firecracker/swagger/firecracker.yaml

```
@@ -943,6 +967,17 @@ definitions:
       vsock:
         $ref: "#/definitions/Vsock"
 
+  Hotplug:
+    type: object
+    description:
+      Defines the configuration for hot plugging.
```

shouldn't `vcpu_count` be required here?

```suggestion
      Defines the configuration for hot plugging.
    required:
      - vcpu_count
```

---

Not necessarily. The way the JSON is being deserialized is 'untagged', so that the resource that is being requested does not have to be identified, it merely matches the provided fields with one of the `HotplugResourceConfig` structs that are part of the `HotplugRequestConfig` enum. This means that if `HotplugMemoryConfig` struct is created, with field `mem_size_mib`, a model request would be:
```
{ "mem_size_mib": 128}
```
Perhaps this should be changed for clarity, to include the type in the request:
```
{ "type": "vcpu", "vcpu_count": 8 }
```
Opinions?

---

I think the way it work now is fine. Adding additional `type` only adds redundant info. But actually I thought deserialization will require request to look like this: 
```json
{ "Vcpu": { "vcpu_count": 1 } }
```

---

There are a few different ways to handle it, see: https://serde.rs/enum-representations.html

---

Ok, nice to know. I would leave this with default settings as we don't use any `tag` modifications in FC. In any case this can be changed in the future.

---

Just to clarify, when you say default settings, are you referring to how you have it above? With Vcpu externally tagged?

---

yes. Also in this case maybe we can rename `vcpu_count` to just `add`, but it is just a thought, as I said, all of this can be changed later if needed.

---

> Not necessarily. The way the JSON is being deserialized is 'untagged', so that the resource that is being requested does not have to be identified, it merely matches the provided fields with one of the HotplugResourceConfig structs that are part of the HotplugRequestConfig enum.

So what happens if the user does not provide anything here? Will it get translated to a `vcpu_count` equal to 0?

---

> > Not necessarily. The way the JSON is being deserialized is 'untagged', so that the resource that is being requested does not have to be identified, it merely matches the provided fields with one of the HotplugResourceConfig structs that are part of the HotplugRequestConfig enum.
> 
> So what happens if the user does not provide anything here? Will it get translated to a `vcpu_count` equal to 0?

I believe serde would throw an error, as it would not be able to match to any of the configs. However, I've just pushed a change to update it to use default settings, as suggested by @ShadowCurse.